packages/grafana-ui/src/components/GraphNG/GraphNG.tsx

```diff
@@ -20,141 +18,124 @@ export const FIXED_UNIT = '__fixed';
 export interface GraphNGProps extends Themeable2 {
   width: number;
   height: number;
-  data: DataFrame[];
-  structureRev?: number; // a number that will change when the data[] structure changes
+  frames: DataFrame[];
+  structureRev?: number; // a number that will change when the frames[] structure changes
   timeRange: TimeRange;
-  legend: VizLegendOptions;
   timeZone: TimeZone;
+  legend: VizLegendOptions;
   fields?: XYFieldMatchers; // default will assume timeseries data
   onLegendClick?: (event: GraphNGLegendEvent) => void;
-  children?: (builder: UPlotConfigBuilder, alignedDataFrame: DataFrame) => React.ReactNode;
+  children?: (builder: UPlotConfigBuilder, alignedFrame: DataFrame) => React.ReactNode;
+
+  prepConfig: (alignedFrame: DataFrame, getTimeRange: () => TimeRange) => UPlotConfigBuilder;
+  propsToDiff?: string[];
+  renderLegend: (config: UPlotConfigBuilder) => React.ReactElement;
```

Comment:
prepConfig +1
propsToDiff +/-1 (maybe a little bit too much ;)  I mean, this should not change that much across GraphNG based visualizations, especially that GraphNG should have a fixed API IMO. In the vis(timeline in particular) you are spreading props on GraphNG that does not belong to it's API - this is usually a source of bugs and confusion in React. Maybe instead this property we could cache the prev props in the consumer and pass a function that would compare the previous and current props of the GraphNG consumer? I mean, this function can be anything in the future that basically returns a bool.

renderLegend - not sure why moving it out? Guess legend will be rendered the same regardless of viz type.

Reply:
timeline will need a different legend -- the legend is for values, not the series

Reply:
>  mean, this should not change that much across GraphNG based visualizations, especially that GraphNG should have a fixed API IMO

But bar chart and timeline need re-init the config based on props not in GraphNG -- any better ideas for how to do that?

Reply:
@ryantxu :
>  Maybe instead this property we could cache the prev props in the consumer and pass a function that would compare the previous and current props of the GraphNG consumer? I mean, this function can be anything in the future that basically returns a bool.

Reply:
If you went the hooks way, there's an excellent and simple `usePrevious` hook that would give you this easily

Reply:
the main motivation for GraphNG is to reduce the boilerplate in consumers and to consolidate the diffing/re-init into one place. if we delegate this diffing to an fn passed in by the parent, we're back to splitting up the invalidation logic and copying the memoization mechanisms, no? unless we're passing these extra props to the uplot plugins directly, i don't see much difference between a graphng API that accepts an external diffing closure, or an array of props that are passed through to components and auto-diffed from a list. doesn't seem much different from some arbitrary `.children` prop that can be completely arbitrary as well :shrug: .

Reply:
lets tackle this point in a follow up PR -- `propsToDiff` is an improvement over our current approach -- but there may be alternative lightweight patterns that are workable